dapr_agents/prompt/agent_prompt.py

```diff
@@ -0,0 +1,225 @@
+            return
+
+        # Always make chat_history available
+        vars_set = set(self.template.input_variables) | {"chat_history"}
+
+        # Inject any attributes the template declares
+        valid_attrs, unused_attrs = self._collect_template_attrs()
+        vars_set |= set(valid_attrs.keys())
+        self.template.input_variables = list(vars_set)
+
+        if unused_attrs:
+            logger.warning(
+                "Agent attributes set but not referenced in prompt_template: "
+                f"{', '.join(unused_attrs)}. Consider adding them to input_variables."
+            )
+
+    def prefill_agent_attributes(self) -> None:
+        """
+        Pre-fill prompt_template with agent attributes if specified in `input_variables`.
+        Uses _collect_template_attrs to avoid duplicate logic and ensure consistency.
+        """
+        if not self.template:
+            return
+
+        # Re-use our helper to split valid vs. unused
+        valid_attrs, unused_attrs = self._collect_template_attrs()
+
+        if unused_attrs:
+            logger.warning(
+                "Agent attributes set but not used in prompt_template: "
+                f"{', '.join(unused_attrs)}. Consider adding them to input_variables."
+            )
+
+        if valid_attrs:
+            self.template = self.template.pre_fill_variables(**valid_attrs)
+            logger.debug(f"Pre-filled template with: {list(valid_attrs.keys())}")
+        else:
+            logger.debug("No prompt_template variables needed pre-filling.")
+
+    def _init(self) -> PromptTemplateBase:
+        """
+        Determines which prompt template to use for the agent:
+        1. If the user supplied one, use it.
+        2. Else if the LLM client already has one, adopt that.
+        3. Else generate a system_prompt and ChatPromptTemplate from agent attributes.
+
+        Returns:
+            PromptTemplateBase: The selected or constructed prompt template.
+        """
+        if self.template:
+            logger.info("Using provided prompt template")
+            return self.template
+
+        # Build from system_prompt or attributes
+        if not self.system_prompt:
+            logger.debug("Constructing system_prompt from attributes")
+            self.system_prompt = self.construct_system_prompt()
+
+        logger.debug("Building ChatPromptTemplate from system_prompt")
+        return self.construct_template()
+
+    def _collect_template_attrs(self) -> tuple[Dict[str, str], List[str]]:
+        """
+        Collect agent attributes for prompt template pre-filling and warn about unused ones.
+        - valid: attributes set on self and declared in prompt_template.input_variables.
+        - unused: attributes set on self but not present in the template.
+        Returns:
+            (valid, unused): Tuple of dict of valid attrs and list of unused attr names.
+        """
+        attrs = ["name", "role", "goal", "instructions"]
```

Comment:
I don't get where this is coming from or if it should be more prominent, like template attributes somewhere.